### PR TITLE
feat(channelGroups): allow re-ordering of groups

### PR DESCRIPTION
**Pull request by FineFindus**
Addresses https://github.com/libre-tube/LibreTube/issues/3558 by allowing channel groups to be rearranged using DnD.

https://github.com/libre-tube/LibreTube/assets/63370021/18d7fbe8-08ff-4feb-a13b-6f18ffeebf1a

This will currently update all groups in the database, this may cause problems with larger sets of groups, in which case it may be worth checking which groups need updating.
Due to the other issues listed in https://github.com/libre-tube/LibreTube/issues/3558 with the Ui, I did not add drag handles.

### CODE REVIEW

**Comment by Bnyro** [comment]
It's not yet self-explaining to users that it's possible to reorder them - I think we should add an indicator for that like it's present in the queue bottom sheet or in the navigation bar order options dialog.

**Comment by FineFindus** [comment]
Channel group bottom sheet with drag handle
I initially didn't add due to the already listed UI issues in https://github.com/libre-tube/LibreTube/issues/3558, as I didn't want to add another icon to the already crowded list of icons.

**Comment by Bnyro** [comment]
> Channel group bottom sheet with drag handle I initially didn't add due to the already listed UI issues in #3558, as I didn't want to add another icon to the already crowded list of icons.
>
I get that, but there's no other way to tell users that they can re-order these.

The drag handle should be at the left though, so that it can be clearly distinguished from the edit and delete icon.

File: app/src/main/java/com/github/libretube/ui/sheets/ChannelGroupsSheet.kt [hunk 1]
```diff
@@ -46,6 +49,35 @@ class ChannelGroupsSheet(
...
+                viewHolder: RecyclerView.ViewHolder,
+                target: RecyclerView.ViewHolder
+            ): Boolean {
+                val from = viewHolder.absoluteAdapterPosition
+                val to = target.absoluteAdapterPosition
+
+                groups.move(from, to)
+                adapter.notifyItemMoved(from, to)
+
+                groups.mapIndexed { index, subscriptionGroup -> subscriptionGroup.index = index }
```
Comment (Isira-Seneviratne) [inline app/src/main/java/com/github/libretube/ui/sheets/ChannelGroupsSheet.kt:67]:
```suggestion
                groups.forEachIndexed { index, subscriptionGroup -> subscriptionGroup.index = index }
```

File: app/src/main/java/com/github/libretube/ui/fragments/SubscriptionsFragment.kt [hunk 2]
```diff
@@ -187,6 +187,7 @@ class SubscriptionsFragment : Fragment() {
         binding.channelGroups.removeAllViews()
 
         binding.channelGroups.addView(binding.chipAll)
+        channelGroups = channelGroups.sortedBy { subscriptionGroup -> subscriptionGroup.index }
```
Comment (Isira-Seneviratne) [inline app/src/main/java/com/github/libretube/ui/fragments/SubscriptionsFragment.kt:190]:
```suggestion
        channelGroups = channelGroups.sortedBy { it.index }
```

**Review by Isira-Seneviratne — APPROVED** [review]
Looks good to me overall, added a couple of small suggestions.

File: app/src/main/java/com/github/libretube/ui/sheets/ChannelGroupsSheet.kt (hunk 1 above)
Comment (FineFindus) [reply app/src/main/java/com/github/libretube/ui/sheets/ChannelGroupsSheet.kt]: I'm not familiar with Kotlin, what's the difference between the two? The documentation wasn't much help.

Comment (Isira-Seneviratne) [reply app/src/main/java/com/github/libretube/ui/sheets/ChannelGroupsSheet.kt]: `mapIndexed` creates a new list with the content specified in the lambda, while `forEachIndexed` just iterates over the items in the list.

**Review by Bnyro — APPROVED** [review]

**Pull request merged**